### PR TITLE
added IAM Role support to hz-cloud

### PR DESCRIPTION
**Pull request by bilalyasar**
closes https://github.com/hazelcast/hazelcast/issues/2630

### CODE REVIEW

File: hazelcast-cloud/src/main/java/com/hazelcast/aws/impl/DescribeInstances.java [hunk 1]
```diff
@@ -59,6 +66,29 @@ public DescribeInstances(AwsConfig awsConfig, String endpoint) {
...
+                    awsConfig.setAccessKey(matcher.group(1));
+                }
+                if (line.contains("SecretAccessKey") && matcher.find()) {
+                    awsConfig.setSecretKey(matcher.group(1));
+                }
+                if (line.contains("Token") && matcher.find()) {
+                    attributes.put("X-Amz-Security-Token", matcher.group(1));
+                }
+            }
+        } catch (IOException io) {
```
Comment (mesutcelik) [inline hazelcast-cloud/src/main/java/com/hazelcast/aws/impl/DescribeInstances.java]: empty catch.. we should fail-fast here.

File: hazelcast-cloud/src/main/java/com/hazelcast/aws/impl/DescribeInstances.java [hunk 2]
```diff
@@ -59,6 +66,29 @@ public DescribeInstances(AwsConfig awsConfig, String endpoint) {
         attributes.put("X-Amz-Expires", "30");
     }
 
+    private void getKeysFromIamRole() {
+        try {
+            String query = "latest/meta-data/iam/security-credentials/" + awsConfig.getIamRole();
+            URL url = new URL("http", "169.254.169.254", -1, query);
```
Comment (mesutcelik) [inline hazelcast-cloud/src/main/java/com/hazelcast/aws/impl/DescribeInstances.java]: you can use following constructor here + pull out IP address into a static final.

```    
public URL(String protocol, String host, String file)
            throws MalformedURLException {
        this(protocol, host, -1, file);
    }
```

**Comment by mesutcelik** [comment]
can you please test with clients both inside and outside the cloud? Also you should add fail fast in case an outside client which tries to connect to a cluster with a role should be disallowed...

File: hazelcast-cloud/src/main/java/com/hazelcast/aws/impl/DescribeInstances.java [hunk 3]
```diff
@@ -59,6 +66,29 @@ public DescribeInstances(AwsConfig awsConfig, String endpoint) {
...
     }
 
+    private void getKeysFromIamRole() {
+        try {
+            String query = "latest/meta-data/iam/security-credentials/" + awsConfig.getIamRole();
+            URL url = new URL("http", "169.254.169.254", -1, query);
+            InputStreamReader is = new InputStreamReader(url.openStream(), "UTF-8");
+            BufferedReader reader = new BufferedReader(is);
+            Pattern pattern = Pattern.compile(": \"(.*?)\",");
+            for (String line; (line = reader.readLine()) != null; ) {
```
Comment (mesutcelik) [inline hazelcast-cloud/src/main/java/com/hazelcast/aws/impl/DescribeInstances.java]: I am sure you can create a static util method which parses response string and returns a map which contains required parameters. That makes testing this part easier...

File: hazelcast-cloud/src/main/java/com/hazelcast/aws/impl/DescribeInstances.java (hunk 1 above)
Comment (bilalyasar) [inline hazelcast-cloud/src/main/java/com/hazelcast/aws/impl/DescribeInstances.java]: thanks, updated

File: hazelcast-cloud/src/main/java/com/hazelcast/aws/impl/DescribeInstances.java (hunk 3 above)
Comment (bilalyasar) [inline hazelcast-cloud/src/main/java/com/hazelcast/aws/impl/DescribeInstances.java]: i have separated them.

**Comment by bilalyasar** [comment]
added new tests, also rebased.

**Pull request merged**